Mitigate BREACH attack

jkowens opened the pull request:
Implement the Rails solution to mitigate BREACH attack. See rails/rails/pull/11729 and rails/rails/pull/16570. Resolves #64.

To actually mitigate the attack forms and meta tags will need to use a masked authenticity token. This can be retrieved by using the following:

```
Rack::Protection::AuthenticityToken.token(session)
```

In a hidden input:

```
<input type="hidden" name="authenticity_token" value="#{Rack::Protection::AuthenticityToken.token(session)}" />
```

Or in a meta tag:

```
<meta name="csrf-token" content="#{Rack::Protection::AuthenticityToken.token(session)" />
```

The `#token` method can also be called on any of the classes that extend `AuthenticityToken` such as `FormToken`, and `RemoteToken`.

Since this uses the same algorithm to mask tokens as Rails, this update will fix compatibility with with Rails >= 4.2.0.

zzak commented:
Haha nice! I was actually hacking together a sloppy version of this myself, glad you got to it first :)

jkowens commented:
Haha sorry! Feel free to go with what you've got started instead.

zzak commented:
@jkowens Well I'd like to merge sinatra/sinatra#1167 first, could you submit your patch upstream after that happens?

My patch isn't ready, and if this works, I'd rather use that then spend more time on it :)

zzak commented on lib/rack/protection/authenticity_token.rb:
```diff
@@ -20,15 +22,91 @@ class AuthenticityToken < Base
       default_options :authenticity_param => 'authenticity_token',
                       :allow_if => nil
 
+      TOKEN_LENGTH = 32
```
This should be default option, which can be configurable, call it `:authenticity_token_length`

jkowens commented:
@zzak Sure thing, I'll get this submitted upstream. I've refactored the code a bit and made the authenticity token length an option. I added a `Rack::Protection::Utils` module to help accomplish that so hopefully that's ok.

zzak commented:
A new module is ok, but I think `Base` is also fine! There's already a `random_string` method which was previously used for generating tokens.

zzak commented:
@jkowens Looking forward to your patch :)
